## Importing Libraries

In [554]:
import pandas as pd
import numpy as np

## Importing Our dataset

In [555]:
movies = pd.read_csv('tmdb_5000_movies.csv')
credits = pd.read_csv('tmdb_5000_credits.csv')

In [556]:
movies.head(3)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466

In [557]:
credits.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [558]:
movies.shape , credits.shape

((4803, 20), (4803, 4))

## Merging both dataframes to make our dataset ready to work

In [559]:
movies = pd.merge(movies,credits)
movies.shape

(4809, 23)

In [560]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

## EDA and Data Preprocessing

### Taking Out Useful Features

In [561]:
movies = movies[['movie_id','genres','keywords','title','cast','crew']]
movies.head()

movie_id                                             genres  \
0     19995  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1       285  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2    206647  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     49026  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     49529  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

### Checking for Missing Values

In [562]:
movies.isnull().sum()

movie_id    0
genres      0
keywords    0
title       0
cast        0
crew        0
dtype: int64

* There are just 3 missing values , we can just **Drop** them

In [563]:
movies.dropna(inplace=True)
movies.shape

(4809, 6)

### Checking "genres" Feature

In [564]:
movies.genres[0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

 Here, we have to abstract **"name"** from each dictionary in genres column gor each movie.
 For this, we first have to change this object type.
* This can be done by using **literal_eval()** function of ast module.
* The **literal_eval** safely evaluate an expression node or a string containing a Python literal or container display. The string or node (a file) provided may only consist of the following Python literal structures: strings, bytes, numbers, tuples, lists, dicts, sets, booleans, and None.

In [565]:
from ast import literal_eval
print(literal_eval(movies.genres[0]))

[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 878, 'name': 'Science Fiction'}]


In [566]:
def get_genres(obj):
    Genres = []
    for itr in literal_eval(obj):
        Genres.append(itr['name']) 
    return Genres 

In [567]:
movies['genres'] = movies['genres'].apply(get_genres)
movies.head()

movie_id                                         genres  \
0     19995  [Action, Adventure, Fantasy, Science Fiction]   
1       285                   [Adventure, Fantasy, Action]   
2    206647                     [Action, Adventure, Crime]   
3     49026               [Action, Crime, Drama, Thriller]   
4     49529           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

### Checking "keyword" Feature

In [568]:
movies.keywords[0]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

Similarly to genres , we have to abstract **"name"** from keywords for each movie

In [569]:
def get_keywords(obj):
    Keywords = []
    for itr in literal_eval(obj):
        Keywords.append(itr['name']) 
    return Keywords 

In [570]:
movies['keywords'] = movies['keywords'].apply(get_genres)
movies.head()

movie_id                                         genres  \
0     19995  [Action, Adventure, Fantasy, Science Fiction]   
1       285                   [Adventure, Fantasy, Action]   
2    206647                     [Action, Adventure, Crime]   
3     49026               [Action, Crime, Drama, Thriller]   
4     49529           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

**'title'** Feature are good to go.
### Checking for "cast" feature

In [571]:
movies.cast[0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In this column, we have various attributes like cast_id, name_of_charachter, credit_id, gender, actor_name.
* In all attributes, only **actor_name** is of much use. So let's abstract **top-5 actors** for each movie

In [572]:
def get_top5_actors(obj):
    Cast = []
    temp = 0
    for itr in literal_eval(obj):
        if temp < 5:
            Cast.append(itr['name'])
        temp+=1
    return Cast 

In [573]:
movies['cast'] = movies['cast'].apply(get_top5_actors)
movies.head()

movie_id                                         genres  \
0     19995  [Action, Adventure, Fantasy, Science Fiction]   
1       285                   [Adventure, Fantasy, Action]   
2    206647                     [Action, Adventure, Crime]   
3     49026               [Action, Crime, Drama, Thriller]   
4     49529           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weave...   
1  [Johnny Depp, Orlando Bloom, Keira Knightley, ...   
2  [Daniel Craig, Christoph Waltz, Léa Seydoux, R...   
3  [Christian Bale, Michael Caine, Gary Oldman, A...   
4  [Taylor Kitsch, Lynn Collins, Samantha Morton,...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [574]:
print(movies.cast[0])
print(len(movies.cast[0]))

['Sam Worthington', 'Zoe Saldana', 'Sigourney Weaver', 'Stephen Lang', 'Michelle Rodriguez']
5


### Checking for "crew" feature

In [575]:
movies.crew[0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In this column, we have various attributes like credit_id, department, gender, job, name.


In [576]:
def get_director(obj):
    Director  = []
    for itr in literal_eval(obj):
        if itr['job'] == 'Director':
            Director.append(itr['name'])
    return Director 

In [577]:
movies['director'] = movies['crew'].apply(get_director)
movies.drop('crew', axis=1, inplace=True)
movies.head()

movie_id                                         genres  \
0     19995  [Action, Adventure, Fantasy, Science Fiction]   
1       285                   [Adventure, Fantasy, Action]   
2    206647                     [Action, Adventure, Crime]   
3     49026               [Action, Crime, Drama, Thriller]   
4     49529           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast             director  
0  [Sam Worthington, Zoe Saldana, Sigourney Weave...      [James Cameron]  
1  [Johnny Depp, Orlando Bloom, Keira Knightley, ...     [Gore Verbinski]  
2  [Daniel Craig, Christoph Waltz, Léa Seydoux, R...         [Sam Mendes]  
3  [Christian Bale, Michael Caine, Gary Oldman, A...  [Christopher Nolan]  
4  [Taylor Kitsch, Lynn Collins, Samantha Morton,...     [Andrew Stanton]

We **converge** genres,keywords,cast,directors into 1 column so we can work on 1 column only to find similarities that will help our model for better recommendations.

But before coverging, we have to perform some **additional preprocessing**.
* In **genres** , we have Science Fiction, which may be considered as 2 seperate identities science and fiction. So we have to convert Science Fiction into something like Science_Fiction or ScienceFiction
* In **cast/director** we have 2 different person with same first name --> Sam Worthington, Sam Mendes. So similarly we have to convert them into SamWorthington, SamMendes

In [578]:
def converge(obj):
    ls = []
    for itr in obj:
        ls.append(itr.replace(" ",""))
    return ls

In [579]:
movies['cast'] = movies['cast'].apply(converge)
movies['director'] = movies['director'].apply(converge)
movies['genres'] = movies['genres'].apply(converge)
movies['keywords'] = movies['keywords'].apply(converge)

In [580]:
movies.head()

movie_id                                        genres  \
0     19995  [Action, Adventure, Fantasy, ScienceFiction]   
1       285                  [Adventure, Fantasy, Action]   
2    206647                    [Action, Adventure, Crime]   
3     49026              [Action, Crime, Drama, Thriller]   
4     49529           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast            director  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver, ...      [JamesCameron]  
1  [JohnnyDepp, OrlandoBloom, KeiraKnightley, Ste...     [GoreVerbinski]  
2  [DanielCraig, ChristophWaltz, LéaSeydoux, Ralp...         [SamMendes]  
3  [ChristianBale, MichaelCaine, GaryOldman, Anne...  [ChristopherNolan]  
4  [TaylorKitsch, LynnCollins, SamanthaMorton, Wi...     [AndrewStanton]

### Converging { genres,  keywords,  cast,  director } into 1 Column --> "features"

In [581]:
movies['features'] = movies['genres'] + movies['keywords'] + movies['cast'] + movies['director']
movies.drop(['genres', 'keywords', 'cast', 'director'], axis=1, inplace=True)
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            features  
0  [Action, Adventure, Fantasy, ScienceFiction, c...  
1  [Adventure, Fantasy, Action, ocean, drugabuse,...  
2  [Action, Adventure, Crime, spy, basedonnovel, ...  
3  [Action, Crime, Drama, Thriller, dccomics, cri...  
4  [Action, Adventure, ScienceFiction, basedonnov...

In [582]:
movies.features[0]

['Action',
 'Adventure',
 'Fantasy',
 'ScienceFiction',
 'cultureclash',
 'future',
 'spacewar',
 'spacecolony',
 'society',
 'spacetravel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alienplanet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'loveaffair',
 'antiwar',
 'powerrelations',
 'mindandsoul',
 '3d',
 'SamWorthington',
 'ZoeSaldana',
 'SigourneyWeaver',
 'StephenLang',
 'MichelleRodriguez',
 'JamesCameron']

* Let's covert everything into 1 paragraph containing all important text; i.e. converting our list to string.

In [583]:
movies['features'] = movies['features'].apply(lambda x: " ".join(x))
movies['features'][0]

'Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver StephenLang MichelleRodriguez JamesCameron'

* Convert everything in feature to small case to avoid any mis-match

In [584]:
movies['features'] = movies['features'].apply(lambda x: x.lower())
movies['features'][0]

'action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver stephenlang michellerodriguez jamescameron'

#### Our dataset seems good. Let's move to building our model 

## Building Recommender System

To build our Recommender System, we use **CountVectorizer()** function to get the similarities between movies by converting all movies into different vectors. Recommended movies are the movies that are closest to the vectors of our watched movie.

In [585]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=3000)

In [586]:
vector = cv.fit_transform(movies['features']).toarray()

In [587]:
cv.get_feature_names()

/opt/homebrew/anaconda3/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['18thcentury',
 '1930s',
 '1940s',
 '1950s',
 '1960s',
 '1970s',
 '1980s',
 '19thcentury',
 '3d',
 'aaroneckhart',
 'aarontaylor',
 'abbiecornish',
 'abigailbreslin',
 'accident',
 'action',
 'actionhero',
 'actor',
 'actress',
 'adambrody',
 'adammckay',
 'adamsandler',
 'adamscott',
 'adamshankman',
 'addiction',
 'adolescence',
 'adoption',
 'adrienbrody',
 'adultanimation',
 'adultery',
 'adventure',
 'affection',
 'afghanistan',
 'africa',
 'africanamerican',
 'aftercreditsstinger',
 'afterlife',
 'agedifference',
 'agent',
 'aging',
 'aidanquinn',
 'aids',
 'airplane',
 'airplanecrash',
 'airport',
 'alanalda',
 'alanarkin',
 'alanrickman',
 'alaska',
 'albertbrooks',
 'alcohol',
 'alcoholabuse',
 'alcoholic',
 'alcoholism',
 'alecbaldwin',
 'alecguinness',
 'alejandrogonzáleziñárritu',
 'alessandronivola',
 'alexanderpayne',
 'alexanderskarsgård',
 'alexapenavega',
 'alexisbledel',
 'alexkendrick',
 'alexpettyfer',
 'alfredhitchcock',
 'alfredmolina',
 'alfrewoodard',
 'alicebr

To find Similarities between movies; we use **Cosine_similarity**. It finds shortest distance between vectors on the basis of angle.

In [588]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vector)
similarity

array([[1.        , 0.12247449, 0.11094004, ..., 0.07071068, 0.        ,
        0.        ],
       [0.12247449, 1.        , 0.1132277 , ..., 0.        , 0.        ,
        0.        ],
       [0.11094004, 0.1132277 , 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.07071068, 0.        , 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [589]:
similarity.shape

(4809, 4809)

* this matrix gives the similarity of all 4809 movies with each movie

Now define a function for recommendation which gives top5 movies based on similarity result

In [590]:
def recommend(movie_name):
    index = movies[movies['title'] == movie_name].index[0]
    distances = sorted(list(enumerate(similarity[index])),reverse=True,key = lambda x: x[1])
    for itr in distances[1:6]:
        print(movies.iloc[itr[0]].title)

### Let's Try our Recommender System

In [591]:
recommend('The Avengers')

Avengers: Age of Ultron
Captain America: Civil War
Captain America: The Winter Soldier
Iron Man 2
Thor: The Dark World


In [592]:
recommend('Men in Black 3')

The Helix... Loaded
Men in Black
Spaced Invaders
Men in Black II
Krrish


In [593]:
recommend('Jurassic World')

Ong Bak 2
Jurassic Park
The Lost World: Jurassic Park
Jurassic Park III
Godzilla 2000


In [604]:
recommend('Twilight')

The Twilight Saga: Breaking Dawn - Part 2
The Twilight Saga: Eclipse
The Twilight Saga: New Moon
Blood and Chocolate
The Mortal Instruments: City of Bones
